src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+        //
+        private void EnsureInternalBufferSize(int curOffset, int addSize)
+        {
+            if (_InternalBuffer == null || _InternalBuffer.Length < addSize + curOffset)
+            {
+                bool wasPinnable = _InternalBufferFromPinnableCache;
+                byte[] saved = _InternalBuffer;
+
+                int newSize = addSize + curOffset;
+                if (newSize <= PinnableReadBufferSize)
+                {
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    {
+                        PinnableBufferCacheEventSource.Log.DebugMessage2("In System.Net._SslStream.EnsureInternalBufferSize IS pinnable", this.GetHashCode(), newSize);
+                    }
+                    _InternalBufferFromPinnableCache = true;
+                    _InternalBuffer = s_PinnableReadBufferCache.AllocateBuffer();
+                }
+                else
+                {
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    {
+                        PinnableBufferCacheEventSource.Log.DebugMessage2("In System.Net._SslStream.EnsureInternalBufferSize NOT pinnable", this.GetHashCode(), newSize);
+                    }
+                    _InternalBufferFromPinnableCache = false;
+                    _InternalBuffer = new byte[newSize];
+                }
+
+                if (saved != null && curOffset != 0)
+                {
+                    Buffer.BlockCopy(saved, 0, _InternalBuffer, 0, curOffset);
+                }
+
+                if (wasPinnable)
+                {
+                    s_PinnableReadBufferCache.FreeBuffer(saved);
+                }
+            }
+            _InternalOffset = curOffset;
+            _InternalBufferCount = curOffset + addSize;
+        }
+
+        //
+        // Validates user parameteres for all Read/Write methods.
+        //
+        private void ValidateParameters(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+
+            if (count > buffer.Length - offset)
+                throw new ArgumentOutOfRangeException("count", SR.net_offset_plus_count);
+        }
+
+        //
+        // Combined sync/async write method. For sync case asyncRequest==null.
+        //
+        private void ProcessWrite(byte[] buffer, int offset, int count, AsyncProtocolRequest asyncRequest)
+        {
+            ValidateParameters(buffer, offset, count);
+
+            if (Interlocked.Exchange(ref _NestedWrite, 1) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, (asyncRequest != null ? "BeginWrite" : "Write"), "write"));
+            }
+
+            bool failed = false;
```

Comment:
Personally, I like `failed = true`, then clearing it in the try-block; that way if multiple catches appear they don't all need to set failed to true.

Not a problem here, just food for thought.